### PR TITLE
fix(bench): copy infrastructure/ after the platform rename

### PR DESCRIPTION
**Pull request by Davidson3556**
Follow-up to #5295. No issue filed: main is red, so I went straight to the fix.

#### Describe the changes you have made in this PR -

`main` is red right now. The `platform` -> `infrastructure` rename (#5295) updated the `benchmark-image.yml` trigger but missed line 63 of `Dockerfile.bench`, so the post-merge image build dies on `COPY platform/`.

The one-line path fix is what turns main green. The rest is a guard test, plus two stale `platform/` references in READMEs that the same rename left behind.

The guard resolves COPY sources against the tracked tree, not the working copy. My first version used `Path.exists()` and happily passed on the broken Dockerfile: a rename leaves untracked `__pycache__` sitting under the old name, and the dockerignore drops `__pycache__`, so Docker sees an empty path while the local check looks fine.

### Demo/Screenshot for feature changes and bug fixes -

The guard reproduces the CI failure on the pre-fix Dockerfile:

```
$ sed -n '63p' tests/benchmarks/cloudopsbench/infra/Dockerfile.bench
COPY platform/ ./platform/

$ uv run python -m pytest tests/benchmarks/cloudopsbench/tests/test_bench_dockerfile_context.py -q
E   AssertionError: Dockerfile.bench copies paths that no longer exist. The bench image build fails post-merge on main, not in PR CI:
E     Dockerfile.bench:63: COPY platform/ -> not tracked in the repo
1 failed
```

And passes with the fix applied:

```
$ sed -n '63p' tests/benchmarks/cloudopsbench/infra/Dockerfile.bench
COPY infrastructure/ ./infrastructure/

$ uv run python -m pytest tests/benchmarks/cloudopsbench/tests/test_bench_dockerfile_context.py -q
1 passed in 1.33s
```

The original failure on main, for reference:

```
#17 [builder 11/15] COPY platform/ ./platform/
#17 ERROR: failed to calculate checksum of ref ...: "/platform": not found
ERROR: failed to build: failed to solve: ... "/platform": not found
```

---

## Code Understanding and AI Usage

**Did you use AI assistance (ChatGPT, Claude, Copilot, etc.) to write any part of this code?**
- [ ] No, I wrote all the code myself
- [x] Yes, I used AI assistance (continue below)

**If you used AI assistance:**
- [x] I have reviewed every single line of the AI-generated code
- [x] I can explain the purpose and logic of each function/component I added
- [x] I have tested edge cases and understand how the code handles them
- [x] I have modified the AI output to follow this project's coding standards and conventions

**Explain your implementation approach:**

Nothing in PR CI builds the bench image, and `Dockerfile.bench` hand-lists source trees instead of copying the whole context. So a rename anywhere in the repo leaves a dangling COPY that nobody sees until it lands on main.

I considered building the image in PR CI. It is slow and needs ECR credentials, which is a lot of machinery for what is really just a question about whether a path exists. A static check answers the same question in about 20ms.

It lives in `tests/benchmarks/` rather than `tests/quality/` because `tests/quality/` is not listed in any shard's `pytest_paths` in `ci.yml`, so a guard there would never run on PRs. That looks like a pre-existing gap, probably worth its own issue. I left it alone here.

`_context_relative_sources()` parses the COPY lines, skips `--from=` (those read from another stage or image, not the build context), drops flags, and treats the last token as the destination. `_tracked_paths()` shells out to `git ls-files` once and expands parent directories into a set so lookups stay O(1) instead of re-scanning per source.

For testing I inverted the fix, confirmed the guard fails with the same error CI gave, then restored it and confirmed it passes. I also ran `tests/core/` (2217 passed, 3 skipped), which `test-scope` selects because of the README edit, and `tests/benchmarks/cloudopsbench/tests/` (215 passed, 18 skipped). ruff and mypy are clean on the new file.

Docker was not available on my machine, so I have not built the image itself. The static check stands in for that. The guard also covers multi-source COPY lines like `COPY pyproject.toml uv.lock ./`, `--chown` flags, and trailing slashes on directory sources.

---

## Checklist before requesting a review
- [ ] I have added proper PR title and linked to the issue (no issue: hotfix for red main, references #5295)
- [x] I have performed a self-review of my code
- [x] **I can explain the purpose of every function, class, and logic block I added**
- [x] I understand why my changes work and have tested them thoroughly
- [x] I have considered potential edge cases and how my code handles them
- [x] If it is a core feature, I have added thorough tests
- [x] My code follows the project's style guidelines and conventions

### CODE REVIEW

File: tests/benchmarks/cloudopsbench/tests/test_bench_dockerfile_context.py [hunk 1]
```diff
@@ -0,0 +1,88 @@
...
+
+def test_bench_dockerfile_copies_only_tracked_paths() -> None:
+    # The benchmark-image workflow builds with ``context: .`` (the repo root).
+    tracked = _tracked_paths()
+    missing = [
+        f"{_DOCKERFILE.name}:{lineno}: COPY {source} -> not tracked in the repo"
+        for lineno, source in _context_relative_sources()
+        if source.rstrip("/") not in tracked
+    ]
+
```
[the remark this answers is not in the document]
Comment (Davidson3556) [reply tests/benchmarks/cloudopsbench/tests/test_bench_dockerfile_context.py]: Good catch, fixed in 8ed94c5.

You are right that tracked-and-excluded is indistinguishable from absent as far as Docker is concerned. `COPY docs/` is the live example: 488 tracked files, dropped from the context by the `docs/` rule, and the old guard waved it through.

The guard now rebuilds the context the way BuildKit does rather than trusting `git ls-files` alone. It walks the tracked files, applies the ignore rules with later rules winning, and keeps whatever survives. Directory patterns are matched against every ancestor so a matched directory drops its subtree and a negation re-admits it, which is what `tests/*` followed by `!tests/benchmarks` depends on.

Verified against three cases: current tree passes, the original `COPY platform/` regression still fails, and a planted `COPY docs/` now fails where it previously passed.
